Add eventlet reactor

justinsb opened the pull request:
Apache licensed, from MagnetoDB project

https://github.com/stackforge/magnetodb/blob/master/magnetodb/common/cassandra/io/eventletreactor.py

aholmberg commented on cassandra/io/eventletreactor.py:
```diff
@@ -0,0 +1,213 @@
+                    self.defunct(err)
+                    return  # leave the read loop
+
+            if self._iobuf.tell():
```
Is there a reason this logic is here, in handle_read instead of left to connection.process_io_buffer?

aholmberg commented on cassandra/io/eventletreactor.py:
```diff
@@ -0,0 +1,213 @@
+try:
+    from cStringIO import StringIO  # ignore flake8 warning: # NOQA
+except ImportError:
+    from StringIO import StringIO  # ignore flake8 warning: # NOQA
```
This doesn't work for python3. Suggest six.StringIO.
However, I'm also not sure why this connection uses an iobuffer distinct from connection._iobuf

ccstolley replied:
Or better, consider io.BytesIO. six.StringIO maps to StringIO.StringIO in python2.x, which is slow. See also https://github.com/datastax/python-driver/commit/38306470bf0fc3f4499bd7e1c2d1eba987325647

aholmberg replied:
Good note. I was actually trying that myself when the second point occurred to me.

aholmberg commented on cassandra/io/eventletreactor.py:
```diff
@@ -0,0 +1,213 @@
+
+    def push(self, data):
+        chunk_size = self.out_buffer_size
+        for i in xrange(0, len(data), chunk_size):
```
Does not work in python3. Suggest six.moves xrange

aholmberg commented:
It looks like the original source may have been updated. Would it be worth incorporating that here?

justinsb commented:
Going to the latest version from magnetodb was a great idea - thanks!  I think they did indeed resolve the issues you identified (and thanks for the code review).

I added the following to the version found there:

* I changed send -> sendall
* I added eventlet_nosetests, which works the same way as gevent_nosetests.  eventlet, like gevent, needs early monkey-patching.

aholmberg commented:
Thanks for the quick turnaround, and for including the test mechanism.
Looking at this now.
